### ISSUE TITLE
Crash with 2**16 lines of input

### ISSUE THREAD

**Issue by aDotInTheVoid** [tree-sitter/tree-sitter#2134]
With the following grammar:

```js
module.exports = grammar({
  name: 'YOUR_LANGUAGE_NAME',

  extras: $ => [
    $.line_comment,
    /\s/,
  ],

  rules: {
    source_file: $ => 'hello',
    line_comment: $ => /#[^\n\r]*[\n\r]*/,
  }
});
```

If I have an input file `big.txt` with conents

```
# Comment
[repeated for 65534 lines]
hello
```

When I try to run `tree-sitter parse` on it I get

```
$ tree-sitter parse ./big.txt
(source_file [0, 0] - [0, 0])
Segmentation fault
```

Adding one more line of `# Comment` gets

```
$ tree-sitter parse ./big.txt
(source_file [1, 0] - [1, 0])
munmap_chunk(): invalid pointer
Aborted
```

And another

```
$ tree-sitter parse ./big.txt
(source_file [0, 0] - [1, 0])
double free or corruption (out)
Aborted
```

If I only have 65534 lines of comments, and then hello (small.txt), then everything works fine.

On `tree-sitter 0.20.7`, Ubuntu 22.10

([This repo](https://github.com/painwacc/tree-sitter-65536), has all the files to reproduce quickly)

**Comment by ahelwer** [comment]
Hmmm, this would seem to indicate some uint16_t is overflowing but the column counter is a uint32_t as seen here: https://github.com/tree-sitter/tree-sitter/blob/457eb2fe761f1e4d0b6346154345007e040fa487/lib/src/point.h#L7

**Comment by aDotInTheVoid** [comment]
Is is possible to run tree sitter under UbSan/Asan? That feels like it would catch this.

**Comment by ahlinc** [comment]
You can download the repo and build a debug version of CLI with `cargo build` and run it with `cargo run -- parse ...`.
Then try to play with it like described in https://users.rust-lang.org/t/howto-sanitize-your-rust-code/9378
May be that would help, I hope.

**Comment by ahelwer** [comment]
I don't have time to do this myself right now but you can use this code to build the language with sanitizer flags; change the build script language name and it should work: https://github.com/tlaplus-community/tree-sitter-tlaplus/tree/main/test/sanitize

**Comment by aDotInTheVoid** [comment]
TLDR: This is fixed in v0.20.8:

[I implemented  sanitized based on TLA+](https://github.com/painwacc/tree-sitter-65536/tree/trunk/sanitize)

### v0.20.7

With v0.20.7, and `-fsanitize=undefined`

```
$ ./runner ../big.txt --quiet
UndefinedBehaviorSanitizer:DEADLYSIGNAL
==5041==ERROR: UndefinedBehaviorSanitizer: SEGV on unknown address 0x7fd36c000000 (pc 0x7fd36d8c0769 bp 0x000000000000 sp 0x7fff58cf7690 T5041)
==5041==The signal is caused by a READ memory access.
    #0 0x7fd36d8c0769 in arena_for_chunk malloc/arena.c:162:47
    #1 0x7fd36d8c0769 in arena_for_chunk malloc/arena.c:160:1
    #2 0x7fd36d8c0769 in cfree malloc/malloc.c:3366:16
    #3 0x559b7fd484f2 in ts_subtree_pool_free /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:148:5
    #4 0x559b7fd3d253 in ts_subtree_release /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:614:7
    #5 0x559b7fd4f20a in ts_tree_delete /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/tree.c:31:3
    #6 0x559b7fceedea in main /home/alona/tmp/tree-sitter-65536/sanitize/runner.cc:35:3
    #7 0x7fd36d844a8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #8 0x7fd36d844b48 in __libc_start_main csu/../csu/libc-start.c:360:3
    #9 0x559b7fcc55d4 in _start (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0x4b5d4) (BuildId: 071c56e717c31291a3210b300d049cebacce75c9)

UndefinedBehaviorSanitizer can not provide additional info.
SUMMARY: UndefinedBehaviorSanitizer: SEGV malloc/arena.c:162:47 in arena_for_chunk
==5041==ABORTING
```

With `-fsanitize=address,undefined`

```
$ ./runner ../big.txt --quiet
=================================================================
==5078==ERROR: AddressSanitizer: attempting double-free on 0x7fa5114df800 in thread T0:
    #0 0x560812700bf2 in __interceptor_free (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0xf0bf2) (BuildId: a1c241fb33ebd9f5c67509081bb78ab5c4b94e34)
    #1 0x5608127e3fd5 in ts_subtree_pool_free /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:148:5
    #2 0x5608127cbd48 in ts_subtree_release /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:614:7
    #3 0x5608127f4fb4 in ts_tree_delete /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/tree.c:31:3
    #4 0x56081273e09c in main /home/alona/tmp/tree-sitter-65536/sanitize/runner.cc:35:3
    #5 0x7fa5135dca8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #6 0x7fa5135dcb48 in __libc_start_main csu/../csu/libc-start.c:360:3
    #7 0x56081267b474 in _start (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0x6b474) (BuildId: a1c241fb33ebd9f5c67509081bb78ab5c4b94e34)

0x7fa5114df800 is located 524288 bytes inside of 655360-byte region [0x7fa51145f800,0x7fa5114ff800)
allocated by thread T0 here:
    #0 0x5608127012c6 in __interceptor_realloc (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0xf12c6) (BuildId: a1c241fb33ebd9f5c67509081bb78ab5c4b94e34)
    #1 0x56081281093c in ts_realloc_default /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/alloc.c:23:18
    #2 0x56081279c06a in array__reserve /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/./array.h:153:24
    #3 0x5608127b8e99 in array__grow /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/./array.h:179:5
    #4 0x5608127b0acc in stack__iter /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/stack.c:384:13
    #5 0x5608127b0acc in ts_stack_pop_all /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/stack.c:570:10
    #6 0x56081277db60 in ts_parser__accept /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/parser.c:936:25
    #7 0x5608127662d9 in ts_parser__advance /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/parser.c:1516:11
    #8 0x5608127605c4 in ts_parser_parse /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/parser.c:1929:14
    #9 0x56081276a9d8 in ts_parser_parse_string_encoding /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/parser.c:1988:10
    #10 0x56081276a6bd in ts_parser_parse_string /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/parser.c:1977:10
    #11 0x56081273df62 in main /home/alona/tmp/tree-sitter-65536/sanitize/runner.cc:26:7
    #12 0x7fa5135dca8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

SUMMARY: AddressSanitizer: double-free (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0xf0bf2) (BuildId: a1c241fb33ebd9f5c67509081bb78ab5c4b94e34) in __interceptor_free
==5078==ABORTING
```

So if their is an overflow, it's not UB, but causing bad data to be free'd

With `-fsanitize=unsigned-integer-overflow`


```
$ ./runner ../big.txt --quiet
lib/src/stack.c:294:39: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:39 in
lib/src/stack.c:294:46: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:46 in
lib/src/stack.c:362:10: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:362:10 in
lib/src/stack.c:335:66: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:335:66 in
lib/src/parser.c:1627:8: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1627:8 in
lib/src/parser.c:1623:70: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1623:70 in
UndefinedBehaviorSanitizer:DEADLYSIGNAL
==5125==ERROR: UndefinedBehaviorSanitizer: SEGV on unknown address 0x7f535c000000 (pc 0x7f535fdf4769 bp 0x000000000000 sp 0x7ffd3b5034e0 T5125)
==5125==The signal is caused by a READ memory access.
    #0 0x7f535fdf4769 in arena_for_chunk malloc/arena.c:162:47
    #1 0x7f535fdf4769 in arena_for_chunk malloc/arena.c:160:1
    #2 0x7f535fdf4769 in cfree malloc/malloc.c:3366:16
    #3 0x557e749d1de3 in ts_subtree_pool_free /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:148:5
    #4 0x557e749ce6e7 in ts_subtree_release /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/subtree.c:614:7
    #5 0x557e749d4aa6 in ts_tree_delete /home/alona/tmp/tree-sitter-65536/sanitize/tree-sitter/lib/src/tree.c:31:3
    #6 0x557e749b7aec in main /home/alona/tmp/tree-sitter-65536/sanitize/runner.cc:35:3
    #7 0x7f535fd78a8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #8 0x7f535fd78b48 in __libc_start_main csu/../csu/libc-start.c:360:3
    #9 0x557e7498e5d4 in _start (/home/alona/tmp/tree-sitter-65536/sanitize/runner+0xc5d4) (BuildId: cd4114781c11fc3f247565bb7d971a6acd57d84e)

UndefinedBehaviorSanitizer can not provide additional info.
SUMMARY: UndefinedBehaviorSanitizer: SEGV malloc/arena.c:162:47 in arena_for_chunk
==5125==ABORTING
```

None of these overflows seem to be the smoking gun, as they're all on `u32`, but the weirdness occours on 2**16 lines.

### v0.20.8

- `-fsanitize=undefined` is fine (no SEGV)
- `-fsanitize=address` is fine
- `-fsanitize=unsigned-integer-overflow` reports errors for both `big.txt` and `small.txt`

```
$ ./runner ../big.txt --quiet
lib/src/stack.c:294:39: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:39 in
lib/src/stack.c:294:46: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:46 in
lib/src/stack.c:362:10: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:362:10 in
lib/src/stack.c:335:66: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:335:66 in
lib/src/parser.c:1631:8: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1631:8 in
lib/src/parser.c:1627:70: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1627:70 in
$ ./runner ../small.txt --quiet
lib/src/stack.c:294:39: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:39 in
lib/src/stack.c:294:46: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'unsigned int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:294:46 in
lib/src/stack.c:362:10: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:362:10 in
lib/src/stack.c:335:66: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/stack.c:335:66 in
lib/src/parser.c:1631:8: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1631:8 in
lib/src/parser.c:1627:70: runtime error: unsigned integer overflow: 4294967295 + 1 cannot be represented in type 'StackVersion' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior lib/src/parser.c:1627:70 in
```

I think these are benign/deliberate, but I'm not sure.

I'm going to close this as fixed, feel free to re-open if you want to add this as a test case.
